projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/intrinsics/c/LLVMSignal.java

```diff
@@ -0,0 +1,425 @@
+public abstract class LLVMSignal extends LLVMFunctionNode {
+
+    // #define SIG_DFL ((__sighandler_t) 0) /* Default action. */
+    private static final LLVMFunctionDescriptor LLVM_SIG_DFL = LLVMFunctionDescriptor.create(0);
+
+    // # define SIG_IGN ((__sighandler_t) 1) /* Ignore signal. */
+    private static final LLVMFunctionDescriptor LLVM_SIG_IGN = LLVMFunctionDescriptor.create(1);
+
+    // #define SIG_ERR ((__sighandler_t) -1) /* Error return. */
+    private static final LLVMFunctionDescriptor LLVM_SIG_ERR = LLVMFunctionDescriptor.create(-1);
+
+    @Specialization
+    public LLVMFunctionDescriptor executeAddress(int signal, LLVMFunctionDescriptor handler) {
+        return setSignalHandler(signal, handler);
+    }
+
+    private static LLVMFunctionDescriptor setSignalHandler(int signalId, LLVMFunctionDescriptor function) {
+        try {
+            Signals decodedSignal = Signals.decode(signalId);
+            return setSignalHandler(decodedSignal.signal(), function);
+        } catch (NoSuchElementException e) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            LLVMLogger.error(e.getMessage());
+            return LLVM_SIG_ERR;
+        }
+    }
+
+    private static Lock globalSignalHandlerLock = new ReentrantLock();
+
+    private static final Map<Integer, LLVMSignalHandler> registeredSignals = new HashMap<>();
+
+    public static int getNumberOfRegisteredSignals() {
+        return registeredSignals.size();
+    }
+
+    @TruffleBoundary
+    private static LLVMFunctionDescriptor setSignalHandler(Signal signal, LLVMFunctionDescriptor function) {
+        int signalId = signal.getNumber();
+        LLVMFunctionDescriptor returnFunction = LLVM_SIG_DFL;
+
+        try {
+            LLVMSignalHandler newSignalHandler = new LLVMSignalHandler(signal, function);
+            synchronized (registeredSignals) {
+                if (registeredSignals.containsKey(signalId)) {
+
+                    LLVMSignalHandler currentFunction = registeredSignals.get(signalId);
+
+                    if (currentFunction.isRunning()) {
+                        returnFunction = currentFunction.getFunction();
+
+                        /*
+                         * the new signal handler already manages this signal, so we can safely
+                         * deactivate the old one.
+                         */
+                        currentFunction.setStopped();
+                    }
+                }
+
+                registeredSignals.put(signalId, newSignalHandler);
+            }
+        } catch (IllegalArgumentException e) {
+            LLVMLogger.error("could not register signal with id " + signalId + " (" + signal + ")");
+            return LLVM_SIG_ERR;
+        }
+
+        return returnFunction;
+    }
+
+    // TODO: stack handling should work without predefined sizes,...
+    private static final long TMP_SIGNAL_STACK_SIZE_KB = 512;
```

Comment:
I think we can just rename this to `SIGNAL_STACK_SIZE_KB` since we already have the TODO here. I do not think that this will ever change anyway.